File: NSPersistentContainer/NSPersistentContainerObjCSample/NSPersistentContainerObjCSample/CoreDataStack.m
//
//  CoreDataStack.m
//  NSPersistentContainerObjCSample
//
//  Created by nikolay.andonov on 10/10/16.
//  Copyright © 2016 nikolay.andonov. All rights reserved.
//

#import "CoreDataStack.h"

@interface CoreDataStack()

@property (strong, nonatomic, readwrite) NSPersistentContainer *persistentContainer;

@end

@implementation CoreDataStack

+ (instancetype)sharedInstance {
    
    static CoreDataStack *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[CoreDataStack alloc] init];
    });
    return sharedInstance;
}

- (NSPersistentContainer *)persistentContainer {
    
    @synchronized (self) {
        if (_persistentContainer) {
            return _persistentContainer;
        }
        
        NSPersistentContainer *persistentContainer = [NSPersistentContainer persistentContainerWithName:@"NSPersistentContainerObjCsSampleModel"];
        [persistentContainer loadPersistentStoresWithCompletionHandler:^(NSPersistentStoreDescription *storeDescription, NSError *error) {
            if (error) {
                NSLog(@"Unresolved Error %@", error);
            }
        }];
        _persistentContainer = persistentContainer;
        return persistentContainer;
    }
}

- (void)saveContext {
    NSManagedObjectContext *context = self.persistentContainer.viewContext;
    if (context.hasChanges) {
        NSError *error = nil;
        [context save:&error];
        if (error) {
            NSLog(@"Unresolved error %@", error);
        }
    }
    
    
}


@end

File: NSQueryGenerationToken/NSQueryGenerationTokenSwiftSample/NSQueryGenerationTokenSwiftSample/CoreDataStack.swift
//
//  CoreDataStack.swift
//  NSPersistentContainerSwiftSample
//
//  Created by nikolay.andonov on 10/9/16.
//  Copyright © 2016 nikolay.andonov. All rights reserved.
//

import CoreData

private let sharedCoreDataStack = CoreDataStack()

class CoreDataStack {
    
    class var sharedInstance: CoreDataStack {
        return sharedCoreDataStack
    }
    
    lazy var persistentContainer: NSPersistentContainer = {
        let container = NSPersistentContainer(name: "NSQueryGenerationTokenModel")
        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
            if let error = error {
                fatalError("Unresolved error \(error))")
            }
        })
        return container
    }()
    
    func saveContext() {
        let context = persistentContainer.viewContext
        if context.hasChanges {
            do {
                try context.save()
            } catch let error as NSError {
                fatalError("Unresolved error \(error)")
            }
        }
    }
}
